ulib/FStar.Matrix.fsti

```diff
@@ -0,0 +1,311 @@
+val matrix_fold_equals_fold_of_seq : 
+  (#c:Type) -> (#eq:CE.equiv c) -> (#m:pos) -> (#n:pos) -> (cm: CE.cm c eq) -> (mx:matrix c m n) 
+  -> Lemma (ensures foldm cm mx `eq.eq` SP.foldm_snoc cm (seq_of_matrix mx)) [SMTPat(foldm cm mx)]
+
+(* A matrix constructed from given generator *)
+val init : (#c:Type) -> (#m:pos) -> (#n: pos) -> (generator: matrix_generator c m n) 
+  -> matrix_of generator 
+
+(* A matrix fold is equal to double foldm_snoc over init-generated seq of seqs *)
+val matrix_fold_equals_fold_of_seq_folds : (#c:Type) -> (#eq: CE.equiv c) -> 
+                                           (#m: pos) -> (#n: pos) ->
+                                           (cm: CE.cm c eq) ->
+                                           (generator: matrix_generator c m n) ->
+  Lemma (ensures foldm cm (init generator) `eq.eq`
+         SP.foldm_snoc cm (SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i))))
+         /\ SP.foldm_snoc cm (seq_of_matrix (init generator)) `eq.eq`
+         SP.foldm_snoc cm (SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i))))         
+         ) 
+
+(* This auxiliary lemma shows that the fold of the last line of a matrix
+   is equal to the corresponding fold of the generator function *)
+ 
+(* This lemma establishes that the fold of a matrix is equal to
+   nested Algebra.CommMonoid.Fold.fold over the matrix generator *)
+val matrix_fold_equals_func_double_fold  : (#c:Type) -> (#eq: CE.equiv c) -> 
+                                           (#m: pos) -> (#n: pos) ->
+                                           (cm: CE.cm c eq) ->
+                                           (generator: matrix_generator c m n) ->
+  Lemma (foldm cm (init generator) `eq.eq` 
+           CF.fold cm 0 (m-1) (fun (i:under m) -> CF.fold cm 0 (n-1) (generator i)))
+ 
+val transposed_matrix_gen (#c:_) (#m:pos) (#n:pos) (generator: matrix_generator c m n) 
+  : (f: matrix_generator c n m { forall i j. f j i == generator i j }) 
+
+val matrix_transpose_is_permutation (#c:_) (#m #n: pos) 
+                                    (generator: matrix_generator c m n)
+  : Lemma (SP.is_permutation (seq_of_matrix (init generator))
+                             (seq_of_matrix (init (transposed_matrix_gen generator)))
+                             (transpose_ji m n))
+                             
+val matrix_fold_equals_fold_of_transpose (#c:_) (#eq:_)
+                                         (#m #n: pos) 
+                                         (cm: CE.cm c eq) 
+                                         (gen: matrix_generator c m n)
+  : Lemma (foldm cm (init gen) `eq.eq`
+           foldm cm (init (transposed_matrix_gen gen)))
+           
+val matrix_eq_fun (#c:_) (#m #n: pos) (eq: CE.equiv c) (ma mb: matrix c m n) : prop
+  
+val matrix_equiv : (#c: Type) ->
+                   (eq:  CE.equiv c) ->
+                   (m: pos) -> (n: pos) ->
+                   CE.equiv (matrix c m n)                   
```

Comment:
This interface says that `matrix_equiv eq m n` is an equivalence relation. That's good. But I don't see any connection in the interface between `matrix_equiv` and `matrix_eq_fun`. And all the lemmas are stated using `matrix_eq_fun`, e.g., `matrix_mul_is_associative'. So, as a client of this interface I can't conclude that the lemma `matrix_mul_is_associative` is actually establishing an equivalence. 

One solution would be:
* don't expose `matrix_eq_fun` in the interface.
* State everything in terms of `(matrix_eq _ _ _ _).eq`

Reply:
Done, in a way that keeps structural integrity the best